### PR TITLE
improve memory usage in sparse.dot

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
operator: dot(csr.T, dense) = rsp (i.e. _backward_dot)
dataset = criteo. _backward_dot(`(1000 x 2M)`.T X `(1000 x 16)`)
env = c4.8xlarge, Ubuntu DL AMI
_backward_dot baseline: 7107.935484 us
_backward_dot improved: 5019.83871 us (1.5x speedup)

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/tensor/dot-inl.h [hunk 1]
```diff
@@ -613,15 +613,15 @@ inline void DotCsrDnsRspImpl(const OpContext& ctx,
...
             ctx.requested[0].get_space_typed<cpu, 1, char>(
             mshadow::Shape1(workspace_size), s);
           dim_t* row_flg = reinterpret_cast<dim_t*>(workspace.dptr_);
-          dim_t* prefix_sum = row_flg + num_rows;
-
-          Fill<false>(s, TBlob(row_flg, mshadow::Shape1(num_rows), cpu::kDevMask), kWriteTo, 0);
-          mxnet_op::Kernel<MarkRowFlgKernel, cpu>::Launch(s, lhs.aux_shape(csr::kIdx)[0], row_flg,
+          // prefix sum array re-uses the row_flg array temp space
+          dim_t* prefix_sum = row_flg;
+          Kernel<op_with_req<set_to_int<0>, kWriteTo>, cpu>::Launch(s, num_rows, row_flg);
```
Comment (ZiyueHuang) [inline src/operator/tensor/dot-inl.h]: Why not use `set_zero`? Or should `set_zero ` in `line 643` also be replaced by `op_with_req<set_to_int<0>, kWriteTo>`?

Comment (ZiyueHuang) [inline src/operator/tensor/dot-inl.h]: Why not use `set_zero`? Or should `set_zero` in line 643 also be replaced by `op_with_req<set_to_int<0>, kWriteTo>`?

Comment (cjolivier01) [reply src/operator/tensor/dot-inl.h]: set_zero is just a 'using' of set_to_int<0>, so it's the same thing.

Comment (cjolivier01) [reply src/operator/tensor/dot-inl.h]: I am curious, why was Fill() removed?  Fill() will be faster on CPU than set_to_int<0> or set_zero in many cases (in almost all but the most extremely large tensor cases, actually).

Comment (eric-haibin-lin) [reply src/operator/tensor/dot-inl.h]: The number of elements in this array is in the order of millions. I tried 2M & 20M, `set_to_int` is several times faster than Fill. What cases did you test? Maybe we should add a switch to `Fill` depending on the size of the tensor

**Comment by piiswrong** [comment]
updates?

**Comment by eric-haibin-lin** [comment]
in sync with master now.

**Pull request merged**